### DISCUSSION TITLE
Raw data unfiltered/filtered is in floating point for recent transmitter firmware

### DISCUSSION THREAD

**Discussion by tynbendad** [NightscoutFoundation/xDrip#2080] [General]
I have an anubis transmitter and cannot read raw data with xdrip+, instead I get "Raw: -0.13" in the bg data table.

I looked at the sensor rx messages using xdrip-js and it looks to me like my transmitter uses floating point for the unfiltered, filtered fields. here is my decoding of the messages I looked at when given controlled input from resistors:
![image](https://user-images.githubusercontent.com/20666529/163507383-684c8a4b-dc52-4b05-bfe7-68a6850237d5.png)

it looks like xdrip+ is using an integer calculation to come up with its unfiltered result (SensorRxMessage.java only reads in integer values, and they are used in https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/G5Model/RawScaling.java) - from my calculations this will not give the same results, e.g., for 400mg/dl the unfiltered will read approx 230mg/dl - this would be fine for calibration if the result was linear but since the exponent is not decoded the integer calculation currently in xdrip+ will not be linear.  

i still don't understand why my raw was -0.13, but i believe there is at least one xdrip+ issue here.

my anubis transmitter is at firmware version 2.27.2.103, and my xdrip+ is at version 2022.04.06 (variant 3).

**Comment by tynbendad** [comment]
btw, I believe for the floating point data 35 is a better multiplier than 34 - using 35 the results for the same resistances are a closer match to the values I see on a 80xxxx transmitter (which uses 34.)

**Reply by tynbendad** [reply]
fyi- here is my version of the xdrip-js implementation for extracting raw data from the message:
[sensor-rx-message.js (1).txt](https://github.com/NightscoutFoundation/xDrip/files/8498834/sensor-rx-message.js.1.txt)

**Reply by tolot27** [reply]
@joshbohmer Any ideas?

**Reply by theunknown86** [reply]
I think he's correct, I can confirm floats are used in places. I would need to spend time to confirm this one specifically but my initial impression is this would be a significant improvement in RAW accuracy over the current implementation

**Reply by tynbendad** [reply]
you can do the math on the values i posted in the first image - at 40MOhms the unfiltered should read about 500mg/dl (this is what my 80xxxx transmitter reads). interpreted as a fp value and multiplied by 35 it matches, while interpreted as xdrip+ currently does you get around 300mg/dl.

**Reply by LoriLori** [reply]
@tynbendad what amount of MOhms could I connect to transmitter contacts that I would get value of  100mg/dl?

**Reply by tynbendad** [reply]
@LoriLori 200MOhms should give you a raw of about 102

**Comment by Navid200** [comment]
What would be the reason for using raw values with a G6?
What would that enable you to do that you cannot already do using the native values coming from the transmitter?

**Reply by tynbendad** [reply]
You can get raw values even when you aren't getting native values.
You can correlate the raw values to the native values when you are getting them.

**Reply by Navid200** [reply]
Can you share an example that you have verified/tested?

For example, when the G6 transmitter issues a confused calibration error and stops producing readings for a while, does it keep sending raw values during that time?

**Comment by tynbendad** [comment]
this is uploaded by logger/xdrip-js, not xdrip+ but it shows what you are asking. raw is always sent by the tx, even for temporary session errors and sensor stopped/warmup...
![redact-166089474-4d02cd67-0efe-458d-be73-1f972f7c0d56](https://user-images.githubusercontent.com/20666529/166094405-3d883f8d-5c58-4cb0-bd29-69de649a2fcd.png)

**Reply by tolot27** [reply]
> raw is always sent by the tx, even for temporary session errors and sensor stopped/warmup...

Does it mean that you can get preliminary raw values during warmup period? That would be an interesting feature.

**Reply by tynbendad** [reply]
yes, absolutely. we've been using raw for warmup, etc. since g5, others have used it since g4. on g6 it only works on 80xxxx,81xxxx, anubis, and maybe a scant few other tx versions.

**Reply by Navid200** [reply]
https://github.com/NightscoutFoundation/xDrip/issues/85

**Comment by tynbendad** [comment]
https://github.com/NightscoutFoundation/xDrip/pull/2094

**Comment by tynbendad** [comment]
this discussion can be closed, @jamorham has made the change in https://github.com/NightscoutFoundation/xDrip/commit/c5eb16243f78a78967b48cdbf0da2a83096b5d6b
